ERCS/erc-7546.md

```diff
@@ -0,0 +1,199 @@
+The Dictionary MUST maintain a mapping of function selectors to implementation addresses.
+
+Here is the reference example written in Solidity:
+```solidity
+mapping(bytes4 functionSelector => address implementation) internal implementations;
+```
+
+Changes to this mapping SHOULD be communicated through an event (or log).
+```solidity
+event ImplementationUpgraded(bytes4 indexed functionSelector, address indexed implementation);
+```
+
+Additionally, the Dictionary SHOULD store an admin account address with rights to modify the mapping.
+```solidity
+address internal admin;
+```
+
+Changes to this admin address SHOULD also be noticed:
+```solidity
+event AdminChanged(address previousAdmin, address newAdmin);
+```
+
+#### 1.2 Functions
+The Dictionary SHOULD provide functions to retrieve and set implementations.
+
+##### 1.2.1 `getImplementation`
+```solidity
+function getImplementation(bytes4 functionSelector) external view returns (address);
+```
+
+##### 1.2.2 `setImplementation`
+```solidity
+function setImplementation(bytes4 functionSelector, address implementation) external;
+```
+The implementation contract MUST implement the function selector to be registered as a key in the mapping (section 3.1). Therefore, it is RECOMMENDED to verify this during registration.
+
+### 2. Proxy Contract
+This contract checks the Dictionary for the associated implementation address based on its function selector and delegatecalls to it.
+
+#### 2.1 Storage & Events
+The Proxy MUST reference the Dictionary address while ensuring a minimal chance of storage clashes. This is because the Dictionary's registered implementations will use this shared storage.
+
+Changes to the Dictionary address SHOULD be noticed:
+```solidity
+event DictionaryUpgraded(address indexed dictionary);
+```
+
+While it is possible to store the Dictionary address in the code area (e.g., using Solidity's immutable or constant), it SHOULD be designed with caution, considering the possibility that if the Dictionary's admin is not the same as the Proxy's admin, the ability to manipulate the implementation could be permanently lost.
+
+#### 2.2 Functions
+All calls to the Proxy MUST be forwarded to the implementation address registered in the Dictionary through the fallback function. Furthermore, as there is a possibility of collisions with function selectors registered in the Dictionary, Proxy SHOULD NOT have external functions.
+
+##### 2.2.1 `constructor`
+It is RECOMMENDED to initialize contracts, other than setting up the Dictionary address, through the Dictionary as well.
+
+##### 2.2.2 `fallback`
+It is RECOMMENDED to utilize well-established and widely-used logic from libraries like Zeppelin or Safe for the logic used in the fallback function. The implementation address to which delegatecall MUST be the returned value queried to the Dictionary (stored in the slot specified in section 2.1) using the `getImplementation(bytes4 functionSelector)` function defined in section 1.2.1.
```

Comment:
Recommendation of using certain library, in my personal opinion, is not a proper specification for it's ill defined and subject to change of circumstance. Consider move to security consideration section.

Reply:
https://github.com/ethereum/ERCs/pull/39/commits/4765ee16a2c7ab34a015c867b819527020d38a9f